finesse/hardware/serial_device.py

```diff
@@ -1,20 +1,43 @@
 """Provides a base class for USB serial devices."""
 from __future__ import annotations
 
+from dataclasses import dataclass
 from typing import Any
 
-from serial import Serial
+from serial import Serial, SerialException
 from serial.tools.list_ports import comports
 
 from finesse.config import BAUDRATES
 from finesse.device_info import DeviceParameter
 
 from .device import AbstractDevice
 
-_serial_ports: list[str] | None = None
+_serial_ports: dict[_USBSerialPortInfo, str] | None = None
```

Comment:
I cannot find when this is populated. You mention it is done once when opening the program.. but where?